packages/react-dom/src/__tests__/ReactTestUtilsAct-test.js

```diff
@@ -0,0 +1,329 @@
+      expect(calledCounter).toBe(5);
+      expect(button.innerHTML).toBe('5');
+
+      document.body.removeChild(container);
+    });
+
+    it('should flush effects recursively', () => {
+      function App() {
+        let [ctr, setCtr] = React.useState(0);
+        React.useEffect(() => {
+          if (ctr < 5) {
+            setCtr(x => x + 1);
+          }
+        });
+        return ctr;
+      }
+
+      const container = document.createElement('div');
+      act(() => {
+        ReactDOM.render(<App />, container);
+      });
+
+      expect(container.innerHTML).toBe('5');
+    });
+
+    it('detects setState being called outside of act(...)', () => {
+      let setValue = null;
+      function App() {
+        let [value, _setValue] = React.useState(0);
+        setValue = _setValue;
+        return (
+          <button id="button" onClick={() => setValue(2)}>
+            {value}
+          </button>
+        );
+      }
+      const container = document.createElement('div');
+      document.body.appendChild(container);
+      let button;
+      act(() => {
+        ReactDOM.render(<App />, container);
+        button = container.querySelector('#button');
+        button.dispatchEvent(new MouseEvent('click', {bubbles: true}));
+      });
+      expect(button.innerHTML).toBe('2');
+      expect(() => setValue(1)).toWarnDev([
+        'An update to App inside a test was not wrapped in act(...).',
+      ]);
+      document.body.removeChild(container);
+    });
+    describe('fake timers', () => {
+      beforeEach(() => {
+        jest.useFakeTimers();
+      });
+      afterEach(() => {
+        jest.useRealTimers();
+      });
+      it('lets a ticker update', () => {
+        function App() {
+          let [toggle, setToggle] = React.useState(0);
+          React.useEffect(() => {
+            let timeout = setTimeout(() => {
+              setToggle(1);
+            }, 200);
+            return () => clearTimeout(timeout);
+          }, []);
+          return toggle;
+        }
+        const container = document.createElement('div');
+
+        act(() => {
+          ReactDOM.render(<App />, container);
+        });
+        act(() => {
+          jest.runAllTimers();
+        });
+
+        expect(container.innerHTML).toBe('1');
+      });
+    });
+
+    it('warns if you return a value inside act', () => {
+      expect(() => act(() => null)).toWarnDev(
+        [
+          'The callback passed to act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+      expect(() => act(() => 123)).toWarnDev(
+        [
+          'The callback passed to act(...) function must return undefined, or a Promise.',
+        ],
+        {withoutStack: true},
+      );
+    });
+
+    it('warns if you try to await an .act call', () => {
+      expect(act(() => {}).then).toWarnDev(
+        [
+          'Do not await the result of calling act(...) with sync logic, it is not a Promise.',
```

Comment:
I don't think we want this anymore, right? Should always await.

Reply:
Is that what we discussed? are we making a breaking change here? (actually asking, I thought we decided not to)

Reply:
I don't think we treat new warnings as a breaking change. We introduce warnings in minors all the time. It's a bit churny, but I think it's worth it. I might not have thought through all the implications though so perhaps we should discuss it during our team meeting.

Reply:
This seems like it would be rather churny given it was prominent in the blog post and others already integrated it into wrapper libraries (like `react-testing-library`). The bigger problem is people would have to convert all their React tests to be async, right? That's not always ideal.

Btw speaking of which, should we reach out to testing libraries to ask them to proxy the returned value?

Reply:
Yes, please let's not force all tests to be async 😬

Reply:
I don't think we need to await all acts. In sync mode, it works fine with sync-y code. if tests queue promises, they can use the async version (and await on that). now, tests written for sync mode _could_ fail in concurrent mode, at which point we have one of two options - rejig the test setup to match fb's so we can resolve promises immediately, or use the async version of act. both are reasonable choices depending on your constraints.

Reply:
It doesn't have anything to do with concurrent mode. The motivation for requiring `await` is it flushes pending promise resolutions.

For example, if you call `jest.advanceTimers(ms)`, that will fire timer events, but the promises won't resolve, which leads to subtle differences in how your component will work in a real environment. (See: https://github.com/facebook/react/pull/14853#discussion_r263034445). Like if a promise handler includes additional setStates, the test will exit before those have a chance to render, potentially leading to false negatives.

Maybe the downsides of using async/await in all tests outweigh the benefits but we shouldn't dismiss it without considering this pitfall.

Reply:
Another way to think about this is that microtasks are another kind of "passive event." Ideally, `act` should always flush microtasks at the end for the same reason it flushes passive events at the end.

Reply:
I'd just like to _only_ have to `await` in my test if my component actually is doing something async (typically with promises). Otherwise I'd like to keep my test sync is all :)

Reply:
I understand the reasons why it's annoying but nobody seems to be addressing my point about dangling effects in promise handlers.

Reply:
I think it's a reasonable compromise to use async `act` in docs and blog posts but not warn for existing usage. People who discover issues or care will learn to convert to async, and eventually the knowledge will propagate.

Reply:
a side effect of making awaiting act mandatory, is people wouldn't be able to make synchronous assertions about their UI (when nesting inside an async act, for example). I think we want to keep this behaviour. marking this bit as resolved, unless y'all think differently.